docs/admin/config.md
---
title: Configuration
summary: Available configuration for spfs and spk
weight: 20
---

## Configuration Files

Both spfs and spk have their own configuration files. The spfs config file can be in json, yaml, toml or ini format. The files are discovered based on extension and loaded from the following locations:

For spfs: `/etc/spfs.toml`, which can be overridden by `~/.config/spfs/spfs.toml`
For spk: `/etc/spk.toml`, which can be overridden by `~/.config/spk/spk.toml`

### Environment Variables

All spfs and spk configuration values can be overridden in the environment. The name of the variable will be the upper-cased name of the config value, separated by underscores, and prefixed with either `SPFS_` or `SPK_`, eg: `SPFS_STORAGE_ROOT`. In cases where the name of the config value contains an underscore, two underscores can be used to disambiguate separators from names, eg: `SPFS_STORAGE__TAG_NAMESPACE`.

### Viewing the Current Configuration

The effective configuration, after all files and environment variables have been
loaded, can be printed as json with the `spfs config` and `spk config` commands.

### SPFS Configuration

```toml
[storage]
# the file system path under which to store all locally created and
# committed spfs data. This should be a local disk with a decent
# amount of free storage space. Spfs supports the same storage
# being used by multiple users, but the directory must then be read
# and writable by anyone.
root = "~/.local/spfs"
# If true, when rendering payloads, allow hard links even if the payload
# is owned by a different user than the current user. Only applies to
# payloads readable by "other". This is often safe for shared read-only
# processes like render machines and artist workstations, but may cause
# permission issues for users that are authoring packages and layers.
allow_payload_sharing_between_users = false
# The tag namespace can be used to separate all spfs tags created in
# this repository from others, essentially segregating the data. This
# can be helpful to set per-user when shared local storage is used so
# that users don't see/edit/delete other's tags
# tag_namespace = "namespace"

# 'origin' is the default remote that should be configured
# for push and pull operations, typically a shared server or
# NFS filesystem. Any number of additional remotes with different
# names can be configured and used with the --remote flag available
# on most commands
[remote.origin]
# Remotes can either be configured by url address or by
# their full broken down configurations, with examples below
# Unless noted all of the extended parameters in the examples below
# can be also added as url parameters in an address.
address = "file:/tmp/spfs-origin"

[remote.filesystem-example]
scheme = "fs"
path = "/path/to/repository"
# create the repository if it does not already exist
create = false
# do not fail at startup if the repository does not exist,
# but only the first time it is used. This should be set
# when create=false but the path might not reliably be
# available and the
lazy = false
# pin the view of the repository to a point in time in the past.
# This only affects the state and availability of tag data
# (and spk packages) rather than object and file data.
# Tags cannot be modified or added to a pinned repository
when = "2020-06-15"
# optional tag namespace under which to store and read all tags
# see storage.tag_namespace for details
# tag_namespace = "namespace"

# the spfs server uses grpc as its communication protocol
[remote.grpc-example]
scheme = "grpc" # or "http2"
address = "my-domain.com:port"
# if true, don't actually attempt to connect until first use
lazy = false
# The global timeout for all requests made in this client
#
# Default is no timeout
timeout_ms = 100
# Maximum message size that the client will accept from the server
#
# Default is 4 Mb
max_decode_message_size_bytes = 1024
# Maximum message size that the client will sent to the server
#
# Default is no limit
max_encode_message_size_bytes = 1024
# see above on pinned repositories
when = "2020-06-15"
# see above on tag namespaces
# tag_namespace = "namespace"

# currently tar repositories must be extracted into a temporary
# folder while in use, and will be saved back into a tarball when
# the program exists. The main purpose for this is to export and
# import spk package archives. Please contact us on github if you
# have a more complex use for tarball repositories and would like
# to discuss a more efficient implementation
[remote.tar-example]
scheme = "tar"
path = "/path/to/archive.tar"

# The proxy repository reads and writes primarily to a single
# repository, but will read from the secondary repositories
# in the case of data missing from the primary one.
[remote.proxy-example]
scheme = "proxy"
# each or the primary and secondary repositories
# can either be the name of another remote in this config
# or a bespoke url address
primary = "origin"
secondary = ["file:/fallback-repository", "tar-example"]


[user]
# The username used when authoring tags.
# defaults to the current username as reported by the system.
name = "unknown"
# The domain of the user is attached to the username in tags,
# eg: user@domain. Defaults to the current hostname.
domain = "my-company.com"

[filesystem]
# The default mount backend to be used for new runtimes.
#
# OverlayFsWithRenders (linux)
#   Renders each layer to a folder on disk, before mounting
#   the whole stack as lower directories in overlayfs. Edits
#   are stored in the overlayfs upper directory.
#
# OverlayFsWithFuse (linux)
#   Mounts a fuse filesystem as the lower directory to
#   overlayfs, using the overlayfs upper directory for edits
#
# FuseOnly (linux)
#   Mounts a fuse filesystem directly
#
# WinFsp (windows)
#   Leverages the win file system protocol system to present
#   dynamic file system entries to runtime processes
#
backend = "OverlayFsWithRenders"
# The "mount" command will be used to mount overlayfs layers when false.
# Direct system calls will be used when true. Defaults to false.
# This option may be removed in the future and behave as if set to "true".
use_mount_syscalls = false
# The named remotes that can be used by the runtime
# file systems to find object data (if possible)
#
# This option is typically only relevant for virtual file
# systems that can perform read-through lookups, such as FUSE.
secondary_repositories = ["origin"]

# OverlayFs related settings.
#
# See kernel.org/doc/html/latest/filesystems/overlayfs.html for details.
#
# It is not recommended to change the default setting, doing so may lead to
# incorrect behavior when using spk.
[filesystem.overlayfs_options]
# Control if the "index" feature is used, if the kernel supports it.
# Defaults to true.
break_hardlinks = true
# Control if the "metadata" feature is used, if the kernel supports it.
# Defaults to true.
metadata_copy_up = true
# Set the redirect_dir mode to use. If not using spk, setting this to "on" may
# improve performance.
#
# Valid values are:
# - on
# - off
# - follow
# - nofollow
#
# Defaults to "follow".
redirect_dir = "follow"

[fuse]
# the number of threads that the fuse filesystem process will create
# in order to operate. More threads may improve performance of the
# mounted filesystem under heavy load, but may also eat resources
# when numerous runtimes are created on the same host at the same time.
# defaults to the lesser of the number of CPUs or 8
worker_threads = 8
# the number of blocking threads used for IO operations in the
# fuse filesystem process. This is a maximum, but blocking threads
# are created and destroyed based on demand.
max_blocking_threads = 512
# Enable a heartbeat between spfs-monitor and spfs-fuse. If spfs-monitor
# stops sending a heartbeat, spfs-fuse will shut down.
enable_heartbeat = true
# How often to send a heartbeat, in seconds
heartbeat_interval_seconds = 60
# How long to allow not receiving a heartbeat before shutting down, in seconds
heartbeat_grace_period_seconds = 300

[monitor]
# the number of threads that the monitor process will create
# in order to operate. This process does very little work so
# there is unlikely to be a need for more.
worker_threads = 2
# the number of blocking threads used for IO operations in the
# runtime monitor process.
max_blocking_threads = 2

# Optional environment variable names to preserve the value when creating an
# spfs runtime.
[environment]
variable_names_to_preserve = ["TMPDIR", "LD_LIBRARY_PATH"]
```

### SPK Configuration

The spk builds on the spfs configuration and adds settings around packaging. Notably, the package repositories are not part of this configuration as spk directly uses the remote repositories configured for spfs.

```toml

# Global metadata labels can be injected whenever a package is built.
#
# This is a list of commands that will output valid json when run.
[[metadata.global]]
command = ["command", "args"]
[[metadata.global]]
command = ["command2", "args"]

[solver]
# If true, the solver will run impossible request checks on the initial requests
check_impossible_initial = false
# If true, the solver will run impossible request checks before
# using a package build to resolve a request
check_impossible_validation = false
# If true, the solver will run impossible request checks to
# use in the build keys for ordering builds during the solve
check_impossible_builds = false
# Increase the solver's verbosity every time this many seconds pass
#
# A solve has taken too long if it runs for more than this
# number of seconds and hasn't found a solution. Setting this
# above zero will increase the verbosity every that many seconds
# the solve runs. If this is zero, the solver's verbosity will
# not increase during a solve.
too_long_seconds = 0
# The maximum verbosity that automatic verbosity increases will
# stop at and not go above.
verbosity_increase_limit = 0
# Maximum number of seconds to let the solver run before halting the solve
#
# Maximum number of seconds to allow a solver to run before
# halting the solve. If this is zero, which is the default, the
# timeout is disabled and the solver will run to completion.
solve_timeout = 0
# Set the threshold of a longer than acceptable solves, in seconds.
long_solve_threshold = 0
# Set the limit for how many of the most frequent errors are
# displayed in solve stats reports
max_frequent_errors = 0
# Comma-separated list of option names to promote to the front of the
# build key order.
build_key_name_order = ""
# Comma-separated list of option names to promote to the front of the
# resolve order.
request_priority_order = ""

# SPK supports the reporting of operational metrics to a
# statsd-compatible server for aggregation.
[statsd]
# Host name of the statsd server
host = ""
# Port number of the statsd server
port = 0
# Prefix to add to all statsd metrics
prefix = ""
# Format to use for statsd metrics, one of:
#   statsd
#   statsd-exporter-librato
format = "statsd"

# SPK supports automated error capture via sentry
[sentry]
# Sentry DSN
dsn = ""
# Sentry environment name
environment = ""
# If set, read the name of the user that sentry will report from this
# environment variable.
#
# This is useful in CI if the CI system has a variable that contains
# the username of the person who triggered the build.
# username_override_var = ""

# SPK supports configuration of these command line defaults
[cli.ls]
# Use all current host's host options by default for filtering in ls
host_filtering = false

# SPK supports some customization of the distro host options.
[host_options.distro_rules.rocky]
# Set a default compat rule for this distro. For example, on Rocky Linux
# packages built on 9.3 would be usable on 9.4.
compat_rule = "x.ab"

# SPK supports using pre-generated repository indexes to speed up solves.
# The index must be created separately. If the index does not exist for a
# repository SPK will continue to solve without it.
#
# Index use can be enabled for each named repository. For example,
# for the 'origin' repository.
[repositories.origin]
use_index = true
# Once enabled, the index settings can be configured for each named repository
[repositories.origin.index]
# SPK supports validating index data before using it. This can be disabled,
# but validating is safer even though it can add some overhead to a solve
# when using an index.
verify_index_before_use = true
# The kind of index format to use for this repository. SPK currently supports
# a flatbuffer based index.
index_kind = "flatb"
# Time to sleep between getting a write lock on the index data,
# for index generation and updates.
lock_sleep_seconds = 6
# Maximum number of times to try to get a write lock on this index's
# data, for index generation and updates.
lock_max_tries = 5
# Name of the configured messaging channel (see MessageChannels below)
# to send index status update messages too.
update_message_channel = "kafka"
# How often index generation and update processing should send
# index update event messages, in milliseconds.
update_event_send_freq_ms = 5000

# SPK can send messages to external messaging systems when packages are:
# published, modified, or removed. Each messaging system must be configured here.
# Currently, only kafka is supported.
#
# To send messages to a kafka system: the name, the brokers, package_updates_topic_name,
# index_updates_topic_name, and the repo_names list must be configured.
# [[messaging]]
# kafka = { brokers = [ name = "kafka", "brokername:port", "otherbrokername:port", ... ], package_updates_topic_name = "spk-package-updates-or-whatever-your-topic-name-is", index_updates_topic_name = "spk-index-updates-or-whatever-your-topic-name-is", repo_names = [ "origin" ] }
#
# SPK supports other optional properties for configuring a kafka messaging system.
# They are shown below with their defaults. Each must be added to the kafka = { ... } 
# table above to configure them for that messaging channel
# The kafka message sending timeout in milliseconds:
#   message_timeout_ms = 5000
# The kafka producer send queue timeout in milliseconds:
#   producer_queue_timeout_ms = 4000
# The timeout while waiting for an index update to complete, in milliseconds
#   index_update_listener_timeout_ms = 10000,
# The kafka message reading session timeout in milliseconds
#   index_update_listener_session_timeout_ms = 10000,
# The index listener's maximum polling interval in milliseconds
#   index_update_listener_max_polling_interval_ms = 10000,
# How far back in time is consider recent enough for messages to be current, in seconds
#   index_update_listener_recent_past_duration_s = 120,
# The timeout for fetching data from the brokers
#   index_update_listener_broker_fetch_timeout_s = 20,
# For example, a kafka messaging channel with all the settings configured
# would look like this:
kafka = { brokers = [ name = "kafka", "brokername:port", "otherbrokername:port", ... ], package_updates_topic_name = "spk-package-updates-or-whatever-your-topic-name-is", index_updates_topic_name = "spk-index-updates-or-whatever-your-topic-name-is", repo_names = [ "origin" ], message_timeout_ms = 5000, producer_queue_timeout_ms = 4000, index_update_listener_timeout_ms = 10000, index_update_listener_session_timeout_ms = 10000, index_update_listener_max_polling_interval_ms = 10000, index_update_listener_recent_past_duration_s = 120, index_update_listener_broker_fetch_timeout_s = 20 }

# SPK can run an index update server, known  as an indexer, When a messaging system
# is configured. An indexer will update a single repository' index when package changes
# happen.
#
# Indexers are configured to use a named message channel (one configured above)
[indexers.kafka]
message_channel_name = "kafka"
# The name of the repository this indexer will monitor.
# An indexer works on a single repository.
repo_name = "origin"
# How often the indexer should send heartbeat messages, to indicate it
# is running, to the messaging system's 'index_updates_topic_name' topic/queue
heartbeat_freq_ms = 60000
# The kafka session timeout for the indexer
session_timeout_ms = 120000
# The maximum polling interval for an indexer for keeping in touch with the kafka brokers
max_polling_interval_ms = 86400000
```


docs/spfs/configuration.md
---
title: Configuration
---

# Configuring SpFS

## Config File

The spfs config file is loaded from the following locations:
- /etc/spfs.conf
- ~/.config/spfs/spfs.conf

### Config Options

```ini
[storage]
# the file system path under which to
# store all locally created and committed
# spfs data. This should be a local disk
# with a decent amount of free storage space
root = "~/.local/spfs"

# 'origin' is the default remote that should be
# configured for push and pull operations, but
# any number of additional remotes with different
# names can be configured and used with the
# --remote flag available on most commands
[remote.<name>]
# address is a URL where the remote is accessible
# the 'file' protocol is a local file path, but
# other protocols may also be available
address = file:/path/to/remote
```


crates/spk-config/src/config.rs
// Copyright (c) Contributors to the SPK project.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/spkenv/spk

use std::collections::HashMap;
use std::sync::{Arc, RwLock};

use config::Environment;
use once_cell::sync::OnceCell;
use serde::{Deserialize, Serialize};
use spfs::Sentry;

use crate::Result;

pub const FLATBUFFER_INDEX_TOKEN: &str = "flatb";

#[cfg(test)]
#[path = "./config_test.rs"]
mod config_test;

static CONFIG: OnceCell<RwLock<Arc<Config>>> = OnceCell::new();

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct Metadata {
    /// List of commands that output global package metatdata in json to be
    /// collected and added to every built package
    pub global: Vec<MetadataCommand>,
}

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct MetadataCommand {
    /// List containing the executable and its arguments
    pub command: Vec<String>,
}

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct Solver {
    /// If true, the solver will run impossible request checks on the initial requests
    pub check_impossible_initial: bool,

    /// If true, the solver will run impossible request checks before
    /// using a package build to resolve a request
    pub check_impossible_validation: bool,

    /// If true, the solver will run impossible request checks to
    /// use in the build keys for ordering builds during the solve
    pub check_impossible_builds: bool,

    /// Increase the solver's verbosity every time this many seconds pass
    ///
    /// A solve has taken too long if it runs for more than this
    /// number of seconds and hasn't found a solution. Setting this
    /// above zero will increase the verbosity every that many seconds
    /// the solve runs. If this is zero, the solver's verbosity will
    /// not increase during a solve.
    pub too_long_seconds: u64,

    /// The maximum verbosity that automatic verbosity increases will
    /// stop at and not go above.
    pub verbosity_increase_limit: u8,

    /// Maximum number of seconds to let the solver run before halting the solve
    ///
    /// Maximum number of seconds to allow a solver to run before
    /// halting the solve. If this is zero, which is the default, the
    /// timeout is disabled and the solver will run to completion.
    pub solve_timeout: u64,

    /// Set the threshold of a longer than acceptable solves, in seconds.
    pub long_solve_threshold: u64,

    /// Set the limit for how many of the most frequent errors are
    /// displayed in solve stats reports
    pub max_frequent_errors: usize,

    /// Comma-separated list of option names to promote to the front of the
    /// build key order.
    pub build_key_name_order: String,

    /// Comma-separated list of option names to promote to the front of the
    /// resolve order.
    pub request_priority_order: String,

    /// Name of the solver, or all, to run when performing a solve
    pub solver_to_run: String,

    /// Name of the solver whose output to show  when multiple solvers are being run.
    pub solver_to_show: String,

    /// Whether to get the solver to use repository indexes, if
    /// available, instead of the repository directly.
    pub use_indexes: bool,

    /// Default setting for indexes, if using indexes is enabled for
    /// the solver.
    pub indexes: Index,
}

/// The settings for a spk repository index
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct Index {
    /// Whether to validate the index data before using it.
    /// Validating is safer but can add overhead at the start of a
    /// solve that uses indexes.
    pub verify_before_use: bool,

    /// What kind of index to use. Only applies if there is more than
    /// one kind of index available for the repository. The default is
    /// 'flatb', a flatbuffers file based index.
    pub kind: String,

    /// Time to sleep between getting a write lock on the index data,
    /// for index generation and updates.
    pub lock_sleep_seconds: u64,

    /// Maximum number of times to try to get a write lock on the
    /// index data, for index generation and updates.
    pub lock_max_tries: u64,

    /// Name of the configured messaging channel (see MessageChannel)
    /// to send index status update messages too.
    pub update_message_channel: String,

    /// How often index generation and update processing should send
    /// index update event messages.
    pub update_event_send_freq_ms: u64,
}

impl Default for Index {
    fn default() -> Self {
        Self {
            // Default to verifying indexes before using them. This is
            // safer but can add some overhead.
            verify_before_use: true,
            kind: String::from(FLATBUFFER_INDEX_TOKEN),
            // Sleeping for 6 seconds between write lock attempts and
            // allowing 5 tries before bailing out, gives a total of
            // about 30 seconds before timing out of getting a lock
            // for index writing.
            lock_sleep_seconds: 6,
            lock_max_tries: 5,
            update_message_channel: String::from("kafka"),
            // 5 seconds in milliseconds
            update_event_send_freq_ms: 5000,
        }
    }
}

/// The settings for a single repository
#[derive(Default, Clone, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct Repository {
    /// Whether to use an index with this repository, if one is
    /// available.
    pub use_index: bool,

    /// Setting for the repositories index, if an index is enabled.
    pub index: Index,
}

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct Statsd {
    /// Host name of the statsd server
    pub host: String,

    /// Port number of the statsd server
    pub port: u16,

    /// Prefix to add to all statsd metrics
    pub prefix: String,

    /// Format to use for statsd metrics
    pub format: String,
}

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct Ls {
    /// Use all current host's host options by default for filtering in ls
    pub host_filtering: bool,
}

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct Cli {
    /// Entries for command line command that have configuration
    pub ls: Ls,
}

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct DistroRule {
    /// The compat rule to set for the distro, e.g., "x.ab"
    pub compat_rule: Option<String>,
}

#[derive(Clone, Default, Debug, Deserialize, Serialize)]
#[serde(default)]
pub struct HostOptions {
    /// A mapping of distro names to recognize and customizations for each.
    pub distro_rules: HashMap<String, DistroRule>,
}

/// Helper for the default kafka message channel name, when not
/// specified in config.
fn default_kafka_channel_name() -> String {
    String::from("kafka")
}

/// Helper for a default kafka message timeout in ms, when not specified in
/// config file.
fn default_kafka_message_timeout_ms() -> u64 {
    // 5 seconds
    5 * 1000
}

/// Helper for a default kafka producer queue timeout in ms, when not specified in
/// config file.
fn default_kafka_producer_queue_timeout_ms() -> u64 {
    // 4 seconds
    4 * 1000
}

/// Helper for a default kafka index update listener's timeout in ms,
/// when not specified in config file.
fn default_kafka_index_update_listener_timeout_ms() -> u64 {
    // 10 seconds
    10 * 1000
}

/// Helper for a default kafka index update listener's session timeout
/// in ms, when not specified in config file.
fn default_kafka_index_update_listener_session_timeout_ms() -> u64 {
    // 10 seconds
    10 * 1000
}

/// Helper for a default kafka index update listener's maximum polling
/// timeout in ms, when not specified in config file.
fn default_kafka_index_update_listener_max_polling_interval_ms() -> u64 {
    // 1 minute
    60 * 1000
}

/// Helper for a default kafka index update listener's recent past
/// duration amount in seconds, when not specified in config file.
fn default_kafka_index_update_listener_recent_past_duration_ms() -> i64 {
    // 2 minutes
    2 * 60 * 1000
}

/// Helper for a default kafka index update listener's broker fetch
/// timeout in seconds, when not specified in config file.
fn default_kafka_index_update_listener_broker_fetch_timeout_ms() -> u64 {
    // 20 seconds
    20 * 1000
}

/// Configuration for using Kafka as a message channel.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct KafkaChannel {
    /// Name of this configured messaging system. Used to distinguish
    /// it from other configured messaging systems.
    #[serde(default = "default_kafka_channel_name")]
    pub name: String,
    /// List of brokers to connect to, specified as "hostname:port"
    pub brokers: Vec<String>,

    /// Topic name to push package updates to
    pub package_updates_topic_name: Option<String>,

    /// Topic name to push index status updates to
    pub index_updates_topic_name: Option<String>,

    /// Names of the SPK repositories that can send package update
    /// messages to the 'package_updates_topic_name'.
    pub repo_names: Vec<String>,

    /// Message sending timeout in milliseconds, defaults to
    /// 5000 ms (5 seconds)
    #[serde(default = "default_kafka_message_timeout_ms")]
    pub message_timeout_ms: u64,

    /// Producer sending queue timeout in milliseconds, defaults to
    /// 4000 ms (4 seconds). This determines how to long retry for if
    /// the producer queue is full.
    #[serde(default = "default_kafka_producer_queue_timeout_ms")]
    pub producer_queue_timeout_ms: u64,

    /// How long an index update message listener should wait after
    /// the last index update message before timing out and assuming
    /// no more updates are coming. Defaults to 10000 ms (10 seconds)
    #[serde(default = "default_kafka_index_update_listener_timeout_ms")]
    pub index_update_listener_timeout_ms: u64,

    /// The kafka session timeout for an index update listener in milliseconds
    #[serde(default = "default_kafka_index_update_listener_session_timeout_ms")]
    pub index_update_listener_session_timeout_ms: u64,

    /// The maximum polling interval for an index update listener in milliseconds
    #[serde(default = "default_kafka_index_update_listener_max_polling_interval_ms")]
    pub index_update_listener_max_polling_interval_ms: u64,

    /// How far back in time counts as recent for an index update
    /// listener when it is trying to workout if an indexer is running
    /// for the repository index it wants to know about, in milliseconds.
    #[serde(default = "default_kafka_index_update_listener_recent_past_duration_ms")]
    pub index_update_listener_recent_past_duration_ms: i64,

    /// The data fetching timeout used by an index update listener
    /// when querying a kafka broker, in milliseconds.
    #[serde(default = "default_kafka_index_update_listener_broker_fetch_timeout_ms")]
    pub index_update_listener_broker_fetch_timeout_ms: u64,
}

/// Types of message channels.
#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum MessageChannel {
    Kafka(KafkaChannel),
}

/// Helper for a default kafka indexer heartbeat frequency in ms when
/// not specified in the config file.
fn default_indexer_heartbeat_freq_ms() -> u64 {
    // 1 minute
    60 * 1000
}

/// Helper for a default kafka indexer session timeout in ms when
/// not specified in the config file.
fn default_indexer_session_timeout_ms() -> u64 {
    // 2 minutes
    120 * 1000
}

/// Helper for a default kafka indexer's maximum polling interval in
/// ms, when not specified in the config file.
fn default_indexer_max_polling_interval_ms() -> u64 {
    // 24 hours
    86400000
}

/// Configuration for an index update server (an indexer)
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Indexer {
    /// Name of the configured message channel to use. The
    /// MessageChannel must have a publish modification topic, an
    /// index status update topic, and at least one repo name
    /// configured.
    pub message_channel_name: String,

    /// SPK repository name to listen for package modifications about.
    /// The name must be present in the configure MessageChannel's
    /// list of repo names. The indexer will ignore package updates
    /// for other repositories, and will only update the named
    /// repository's index,
    pub repo_name: String,

    /// How frequently this indexer should send heartbeat messages to
    /// the index status channel, when not updating an index. Defaults
    /// to 60000 ms (1 minute). When updating an index it will send
    /// index status messages more frequently.
    #[serde(default = "default_indexer_heartbeat_freq_ms")]
    pub heartbeat_freq_ms: u64,

    /// Indexer's kafka session timeout in milliseconds
    #[serde(default = "default_indexer_session_timeout_ms")]
    pub session_timeout_ms: u64,

    /// Indexer's kafka broker maximum polling interval in milliseconds
    #[serde(default = "default_indexer_max_polling_interval_ms")]
    pub max_polling_interval_ms: u64,

    /// Name of the metric to update whenever the indexer sends an
    /// index event message (including heartbeats). With suitable
    /// external systems, this can be used to monitor an indexer's
    /// workload and whether it is healthy. If it is not set, or
    /// statsd is not configured, the metric will not be updated.
    #[serde(default)]
    pub index_event_metric_name: Option<String>,

    /// Name of the metric to update whenever an indexer starts
    /// updating an index. With suitable external systems, this can be
    /// used to see how often an index is updated. If it is not set,
    /// or statsd is not configured, the metric will not be updated.
    #[serde(default)]
    pub index_update_start_metric_name: Option<String>,
}

/// Configuration values for spk.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(default)]
pub struct Config {
    // These sub-types should aim to only have one level of
    // values within them, otherwise they become impossible to address
    // with environment variables.
    pub sentry: Sentry,
    pub solver: Solver,
    pub repositories: HashMap<String, Repository>,
    pub statsd: Statsd,
    pub metadata: Metadata,
    pub cli: Cli,
    pub host_options: HostOptions,
    pub messaging: Vec<MessageChannel>,
    pub indexers: HashMap<String, Indexer>,
}

impl Config {
    /// Get the current loaded config, loading it if needed
    pub fn current() -> Result<Arc<Self>> {
        get_config()
    }

    /// Load the config from disk, even if it's already been loaded before
    pub fn load() -> Result<Self> {
        load_config()
    }

    /// Make this config the current global one
    pub fn make_current(self) -> Result<Arc<Self>> {
        // Note we don't know if we won the race to set the value here,
        // so we still need to try to update it.
        let config = CONFIG.get_or_try_init(|| -> Result<RwLock<Arc<Config>>> {
            Ok(RwLock::new(Arc::new(self.clone())))
        })?;

        let mut lock = config
            .write()
            .map_err(|err| crate::Error::LockPoisonedWrite(err.to_string()))?;
        *Arc::make_mut(&mut lock) = self;
        Ok(Arc::clone(&lock))
    }
}

/// Get the current spk config, fetching it from disk if needed.
pub fn get_config() -> Result<Arc<Config>> {
    let config = CONFIG.get_or_try_init(|| -> Result<RwLock<Arc<Config>>> {
        Ok(RwLock::new(Arc::new(load_config()?)))
    })?;
    let lock = config
        .read()
        .map_err(|err| crate::Error::LockPoisonedRead(err.to_string()))?;
    Ok(Arc::clone(&*lock))
}

/// Get the current index config for the given repo name
pub fn get_index_config(repo_name: &str) -> Index {
    match get_config() {
        Ok(config) => {
            if let Some(repo_config) = config.repositories.get(repo_name) {
                repo_config.index.clone()
            } else {
                Index::default()
            }
        }
        Err(err) => {
            tracing::warn!("Unable to read spk config file, using Index defaults, due to: {err}");
            Index::default()
        }
    }
}

/// Get the index update server (indexer) config with the given name,
/// if one exits.
pub fn get_indexer_config(indexer_name: &str) -> Option<Indexer> {
    match get_config() {
        Ok(config) => config.indexers.get(indexer_name).cloned(),
        Err(err) => {
            tracing::warn!("Unable to read spk config file to get indexers config, due to: {err}");
            None
        }
    }
}

/// Load the spk configuration from disk, even if it has already been loaded.
///
/// This includes the default, user, and system configurations (if they exist).
pub fn load_config() -> Result<Config> {
    use config::{Config as RawConfig, File};

    const USER_CONFIG_BASE: &str = "spk/spk";
    let user_config = dirs::config_local_dir()
        .map(|config| config.join(USER_CONFIG_BASE))
        .ok_or_else(|| {
            crate::Error::Config(config::ConfigError::NotFound(
                "User config area could not be found, this platform may not be supported".into(),
            ))
        })?;

    let config = RawConfig::builder()
        // the system config can also be in any support format: toml, yaml, json, ini, etc
        .add_source(File::with_name("/etc/spk").required(false))
        // the user config can also be in any support format: toml, yaml, json, ini, etc
        .add_source(File::with_name(&format!("{}", user_config.display())).required(false))
        // Note: if a var using single underscores is set, it will have precedence
        .add_source(
            Environment::with_prefix("SPK")
                .prefix_separator("_")
                .separator("__"),
        )
        // for backwards compatibility with vars not using double underscores
        .add_source(Environment::with_prefix("SPK").separator("_"))
        .build()?;

    Ok(Config::deserialize(config)?)
}
